sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1013 @@
+from ..utils.random import check_random_state
+from ..utils.multiclass import check_classification_targets
+from ..utils.validation import check_is_fitted, check_array, check_X_y, \
+    check_consistent_length
+
+
+class LargeMarginNearestNeighbor(BaseEstimator, TransformerMixin):
+    """Distance Metric Learning for Large Margin Classification.
+
+    Parameters
+    ----------
+    n_features_out : int, optional (default=None)
+        Preferred dimensionality of the embedding.
+        If None it is inferred from ``init``.
+
+    init : string or numpy array, optional (default='pca')
+        Initialization of linear transformation. Possible options are 'pca',
+        'identity' and a numpy array of shape (n_features_out, n_features).
+
+    warm_start : bool, optional, (default=False)
+        If True and :meth:`fit` has been called before, the solution of the
+        previous call to :meth:`fit` is used as the initial linear
+        transformation (``init`` is ignored).
+
+    n_neighbors : int, optional (default=3)
+        Number of neighbors to use as target neighbors for each sample.
+
+    algorithm : {'auto', 'ball_tree', 'kd_tree', 'brute'}, optional
+        Algorithm used to compute the target neighbors:
+
+        - 'ball_tree' will use :class:`BallTree`
+        - 'kd_tree' will use :class:`KDTree`
+        - 'brute' will use a brute-force search.
+        - 'auto' will attempt to decide the most appropriate algorithm
+          based on the values passed to :meth:`fit` method.
+
+        Note: fitting on sparse input will override the setting of
+        this parameter, using brute force.
+
+    targets : array, shape (n_samples, n_neighbors), optional (default=None)
+        The target neighbors of each sample, namely the `n_neighbors` nearest
+        neighbors that belong to the same class. If None (default), the target
+        neighbors will be computed on the fly using the euclidean metric.
+
+    max_constraints : int, optional (default=500000)
+        Maximum number of constraints to enforce per iteration.
+
+    use_sparse : bool, optional (default=True)
+        Whether to use a sparse matrix (default) or a dense matrix for the
+        impostor-pairs storage. Using a sparse matrix, the (squared)
+        distance to impostors is computed twice, but it is faster for larger
+        data sets than using a dense matrix. With a dense matrix, the unique
+        impostor pairs have to be identified explicitly.
+
+    max_iter : int, optional (default=50)
+        Maximum number of iterations in the optimization.
+
+    tol : float, optional (default=1e-5)
+        Convergence tolerance for the optimization.
+
+    max_corrections : int, optional (default=100)
+        The maximum number of variable metric corrections
+        used to define the limited memory matrix. (The limited memory BFGS
+        method does not store the full hessian but uses this many terms in an
+        approximation to it.)
```

Comment:
is it really necessary to expose this parameter? 100 memory steps for L-BFGS seems quite large for me.

Reply:
is it really necessary? Please show me with an experimental benchmark that it should be tuned to be more than marginally faster.

Reply:
You're right, I tried the default value (10) on MNIST, and got practically no speed difference.

Reply:
this is one dataset. Is it replicable across datasets?